File: lr.py
import numpy as np
import math as mt
import scipy as sp
import statsmodels.formula.api as smf
import matplotlib.pyplot as plt
import pandas as pd

# Data simulation.
np.random.seed(666)
n = 30
X = np.column_stack((np.ones(n), np.random.normal(15, 5, n)))
b = np.array((5, 30)).T
p = X.shape[1]
sigma = 30
y = X @ b + np.array((np.random.normal(0, sigma, n))).T
x_tilde = np.arange(mt.floor(min(X[:, 1])), mt.ceil(max( X[:, 1])) + 1, 1)
X_tilde = np.column_stack((np.ones(len(x_tilde)), x_tilde))

# Classical linear regression.
fit = smf.OLS(y, X).fit()
print(fit.summary())
dir(fit)
y_hat = fit.predict(X_tilde)
sigma_hat = np.sqrt(fit.mse_resid)
alph = 0.05

# Graphical representation.
def linreg(y, X, y_hat, sigma_hat, X_tilde):
    invXtX = np.linalg.inv(X.T @ X)
    np.sqrt(np.diag(X_tilde @ invXtX @ X_tilde.T))
    plt.scatter(X[:, 1], y, c = '#000000', alpha = 0.5)
    plt.plot(X_tilde[:, 1], y_hat, c = '#FF0000')
    plt.plot(X_tilde[:, 1], y_hat + 
             sp.stats.t.ppf(0.975, n - p) * sigma_hat * np.sqrt(np.diag(X_tilde @ invXtX @ X_tilde.T)), 
             c = "#00FF00")
    plt.plot(X_tilde[:, 1], y_hat - 
             sp.stats.t.ppf(0.975, n - p) * sigma_hat * np.sqrt(np.diag(X_tilde @ invXtX @ X_tilde.T)), 
             c = "#00FF00")
    plt.plot(X_tilde[:, 1], y_hat + 
             sp.stats.t.ppf(0.975, n - p) * sigma_hat * np.sqrt(1 + np.diag(X_tilde @ invXtX @ X_tilde.T)), 
             c = "#0000FF")
    plt.plot(X_tilde[:, 1], y_hat - 
             sp.stats.t.ppf(0.975, n - p) * sigma_hat * np.sqrt(1 + np.diag(X_tilde @ invXtX @ X_tilde.T)), 
             c = "#0000FF")
    plt.xlabel("x")
    plt.ylabel("y")

linreg(y, X, y_hat, sigma_hat, X_tilde)

# Sample from a multivariate t distribution.
def random_mv_t(df, Sigma, n):
    p = Sigma.shape[1]
    den = np.sqrt(np.random.chisquare(df, n)/df)
    num = np.random.multivariate_normal(np.zeros(p), Sigma, n)
    return num/den[:, None]

# Bayesian linear regression.
def bayes_lm(y, X, X_tilde = None,
             beta_0 = None, Sigma = None, nu_0 = None, sigma2_0 = None,
             size = 10000, alph = 0.05, seed = 666):
    
    # Default values.
    np.random.seed(seed)
    n = len(y)
    p = X.shape[1]
    if beta_0 is None:
        beta_0 = np.zeros(p).T
    if Sigma is None:
        Sigma = 1000000 * np.diag(np.ones(p))
    if nu_0 is None:
        nu_0 = 1
    if sigma2_0 is None:
        sigma2_0 = 1
    
    # Fixed values and auxiliar variables.
    V = X.T @ X
    beta_hat = np.linalg.inv(V) @ X.T @ y
    Sigma_n = np.linalg.inv(np.linalg.inv(Sigma) + V)
    eig_Sigma_n = np.linalg.eigh(Sigma_n)
    eig_Sigma_n = eig_Sigma_n[1] @ np.diag(np.sqrt(eig_Sigma_n[0]))
    beta_n = Sigma_n @ (X.T @ y + Sigma @ beta_0)
    S2 = sum((y - X @ beta_hat)**2)/(n - p)
    
    # Simulation of sigma2 and beta.
    sigma2_sample = (nu_0 * sigma2_0 + (n - p) * S2)/np.random.chisquare(n - p, size)
    beta_sample = np.random.normal(0, 1, size * p)
    beta_sample.shape = (p, size)
    beta_sample = eig_Sigma_n @ (beta_sample * np.sqrt(sigma2_sample)) + beta_n[:, None]
    
    # Prediction.
    if X_tilde is None:
        y_hat = X @ beta_sample
    else:
        m = X_tilde.shape[0]
        sim = random_mv_t(n - p, np.diag(np.ones(m)), size)
        delta = X_tilde @ beta_hat
        y_hat = sim.T * np.sqrt(S2) + delta[:, None]
    
    # Results.
    results = np.column_stack((beta_sample.mean(axis = 1), 
                               np.percentile(beta_sample, axis = 1, q = alph/2 * 100), 
                               np.percentile(beta_sample, axis = 1, q = 100 - alph/2 * 100)))
    sig = np.array((np.mean(sigma2_sample), 
                    np.percentile(sigma2_sample, q = alph/2 * 100), 
                    np.percentile(sigma2_sample, q = 100 - alph/2 * 100)))
    results = np.vstack((results, sig))
    param = ["beta_" + str(i) for i in range(p)]
    param.append("sigma2")
    results = pd.DataFrame({"mean": results[:, 0],
                           "lower": results[:, 1],
                           "upper": results[:, 2]})
    results.index = param
    y_hat = np.column_stack((y_hat.mean(axis = 1), 
                             np.percentile(y_hat, axis = 1, q = alph/2 * 100),
                             np.percentile(y_hat, axis = 1, q = 100 - alph/2 * 100)))
    y_hat = pd.DataFrame({"mean": y_hat[:, 0],
                          "lower": y_hat[:, 1],
                          "upper": y_hat[:, 2]})
    return {"results": results,
            "beta_sample": beta_sample.T,
            "sigma2_sample": sigma2_sample,
            "y_hat": y_hat}

fit1 = bayes_lm(y, X)    
fit2 = bayes_lm(y, X, X_tilde)
fit2["results"]

def bayes_linreg(y, x, y_hat, y_hat2, x_tilde):
    plt.scatter(x, y, c = "#000000", alpha = 0.5)
    plt.plot(x_tilde, y_hat["mean"], c = "#FF0000")
    plt.plot(x, y_hat2["lower"], c = "#00FF00")
    plt.plot(x, y_hat2["upper"], c = "#00FF00")
    plt.plot(x_tilde, y_hat["lower"], c = "#0000FF")
    plt.plot(x_tilde, y_hat["upper"], c = "#0000FF")
    plt.xlabel("x")
    plt.ylabel("y")

y_hat = fit2["y_hat"]
y_hat2 = fit1["y_hat"]
x = X[:,1]
x_tilde = X_tilde[:,1]

bayes_linreg(y, x, y_hat, y_hat2, x_tilde)

File: lr.R
library(ggplot2) # Graphics.
library(mvtnorm) # Multivariate normal and multivariate t.

# Data simulation.
set.seed(69)
n <- 30
X <- cbind(1, rnorm(n, 15, 5))
p <- ncol(X)
b <- c(3, 30)
sigma <- 30
y <- X %*% b + rnorm(n, 0, sigma)
train_data <- as.data.frame(cbind(y, X[, 2]))
names(train_data) <- c("y", "x")
X_tilde <- as.data.frame(cbind(y = NA, 
                               x = seq(floor(min(X[, 2])), ceiling(max(X[, 2])))))

# Classical linear regression.
fit <- lm(y ~ x, train_data)
sum.fit <- summary(fit)
sum.fit
names(fit)
y_hat_ci <- predict(fit, X_tilde, interval = "confidence")
y_hat_pi <- predict(fit, X_tilde, interval = "prediction")

# Graphical representation of the classical linear regression.
ggplot() + geom_point(aes(y = y, x = X[, 2]), alpha = 0.5) +
  geom_line(aes(y = y_hat_ci[, "fit"], 
                x = X_tilde[, 2]), 
            colour = "#FF0000") +
  geom_line(aes(y = y_hat_ci[, "lwr"],
                x = X_tilde[, 2]), 
            colour = "#00FF00") +
  geom_line(aes(y = y_hat_ci[, "upr"],
                x = X_tilde[, 2]), 
            colour = "#00FF00") +
  geom_line(aes(y = y_hat_pi[, "lwr"],
                x = X_tilde[, 2]), 
            colour = "#0000FF") +
  geom_line(aes(y = y_hat_pi[, "upr"],
                x = X_tilde[, 2]), 
            colour = "#0000FF") +
  scale_x_continuous("x", breaks = seq(5, 30, 5)) +
  scale_y_continuous("y", breaks = seq(0, 1000, 200), limits = c(0, 1000))

# Bayesian linear regression.
bayes_lm <- function(y, X, X_tilde = NULL,
                     beta_0 = NULL, Sigma = NULL, nu_0 = NULL, sigma2_0 = NULL, 
                     size = 10000, alph = 0.05, seed = 666) {
  
  # Default values.
  set.seed(seed)
  p <- ncol(X)
  n <- length(y)
  if (is.null(beta_0)) {
    beta_0 <- rep(0, p)
  }
  if (is.null(Sigma)) {
    Sigma <- diag(rep(100000, p))
  }
  if (is.null(nu_0)) {
    nu_0 <- 1
  }
  if (is.null(sigma2_0)) {
    sigma2_0 <- 1
  }
  
  # Fixed values and auxiliar variables.
  V <- t(X) %*% X
  beta_hat <- solve(V) %*% t(X) %*% y
  Sigma_n <- solve(solve(Sigma) + V)
  eig_Sigma_n <- eigen(Sigma_n)
  eig_Sigma_n <- eig_Sigma_n$vectors %*% diag(sqrt(eig_Sigma_n$values)) # spectral decomposition.
  beta_n <- Sigma_n %*% (t(X) %*% y + Sigma %*% beta_0)
  S2 <- sum((y - X %*% beta_hat)^2)/(n - p)
  
  # Simulation of sigma2 and beta.
  sigma2_sample <- (nu_0 * sigma2_0 + (n - p) * S2)/rchisq(size, n - p)
  beta_sample <- t(matrix(rnorm(p * size), ncol = p) * sqrt(sigma2_sample))
  beta_sample <- eig_Sigma_n %*% beta_sample + as.vector(beta_n)
  
  # Prediction.
  if (is.null(X_tilde)) {
    y_hat <- X %*% beta_sample
  } else {
    m <- nrow(X_tilde)
    y_hat <- t(rmvt(size, sigma = diag(rep(1, m)), df = n - p) * sqrt(S2)) +
      c(X_tilde %*% beta_hat)
  }
  
  # Results.
  results <- cbind(rowMeans(beta_sample), 
                   t(apply(beta_sample, 1, quantile, probs = c(alph/2, 1 - alph/2))))
  results <- rbind(results, c(mean(sigma2_sample), 
                              quantile(sigma2_sample, probs = c(alph/2, 1 - alph/2))))
  results <- as.data.frame(results)
  results <- cbind(c(paste0("beta_", 0:(p - 1)), "sigma2"), results)
  names(results) <- c("param", "mean", "lower", "upper")
  y_hat <- cbind(seq(1, nrow(y_hat)), rowMeans(y_hat),
                  t(apply(y_hat, 1, quantile, probs = c(alph/2, 1 - alph/2))))
  y_hat <- as.data.frame(y_hat)
  names(y_hat) <- c("obs", "mean", "lower", "upper")
  return(list(results = results,
              beta_sample = beta_sample,
              sigma2_sample = sigma2_sample,
              y_hat = y_hat))
}

X_tilde <- cbind(1, x = seq(floor(min(X[, 2])), ceiling(max(X[, 2]))))
fit1 <- bayes_lm(y, X)
fit2 <- bayes_lm(y, X, X_tilde)
fit2$results

# Graphical representation of the Bayesian linear regression.
ggplot() + geom_point(aes(y = y, x = X[, 2]), alpha = 0.5) +
  geom_line(aes(y = fit2$y_hat[, "mean"], 
                x = X_tilde[, 2]), 
            colour = "#FF0000") +
  geom_line(aes(y = fit1$y_hat[, "lower"],
                x = X[, 2]), 
            colour = "#00FF00") +
  geom_line(aes(y = fit1$y_hat[, "upper"],
                x = X[, 2]), 
            colour = "#00FF00") +
  geom_line(aes(y = fit2$y_hat[, "lower"],
                x = X_tilde[, 2]), 
            colour = "#0000FF") +
  geom_line(aes(y = fit2$y_hat[, "upper"],
                x = X_tilde[, 2]), 
            colour = "#0000FF") +
  scale_x_continuous("x", breaks = seq(5, 30, 5)) +
  scale_y_continuous("y", breaks = seq(0, 1000, 200), limits = c(0, 1000))
